tokio/src/task/task_set.rs

```diff
@@ -0,0 +1,222 @@
+    pub fn spawn_local_on<F>(&mut self, task: F, local_set: &LocalSet)
+    where
+        F: Future<Output = T>,
+        F: 'static,
+    {
+        self.insert(local_set.spawn_local(task));
+    }
+
+    fn insert(&mut self, jh: JoinHandle<T>) {
+        let mut entry = self.inner.insert_idle(jh);
+
+        // Set the waker that is notified when the task completes.
+        entry.with_value_and_context(|jh, ctx| jh.set_join_waker(ctx.waker()));
+    }
+
+    /// Wait until one of the tasks in the set completes and returns its output.
+    ///
+    /// Returns `None` if the set is empty.
+    ///
+    /// # Cancel Safety
+    ///
+    /// This method is cancel safe. If `join_one` is used as the event in a `tokio::select!`
+    /// statement and some other branch completes first, it is guaranteed that no tasks were
+    /// removed from this `TaskSet`.
+    pub async fn join_one(&mut self) -> Result<Option<T>, JoinError> {
+        crate::future::poll_fn(|cx| self.poll_join_one(cx)).await
+    }
+
+    /// Abort all tasks on this `TaskSet`.
+    ///
+    /// This does not remove the tasks from the `TaskSet`. To wait for the tasks to complete
+    /// cancellation, you should call `join_one` in a loop until the `TaskSet` is empty.
+    pub fn abort_all(&mut self) {
+        self.inner.for_each(|jh| jh.abort());
+    }
+
+    /// Remove all tasks from this `TaskSet` without aborting them.
+    ///
+    /// The tasks removed by this call will continue to run in the background even if the `TaskSet`
+    /// is dropped.
+    pub fn detach_all(&mut self) {
+        self.inner.drain(drop);
+    }
+
+    /// Poll for one of the tasks in the set to complete.
+    ///
+    /// If this returns `Poll::Ready(Some(_))`, then the task that completed is removed from the
+    /// set.
+    ///
+    /// When the method returns `Poll::Pending`, the `Waker` in the provided `Context` is scheduled
+    /// to receive a wakeup when a task in the `TaskSet` completes. Note that on multiple calls to
+    /// `poll_join_one`, only the `Waker` from the `Context` passed to the most recent call is
+    /// scheduled to receive a wakeup.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    ///  * `Poll::Pending` if the `TaskSet` is not empty but there is no task whose output is
+    ///     available right now.
+    ///  * `Poll::Ready(Ok(Some(value)))` if one of the tasks in this `TaskSet` has completed. The
+    ///    `value` is the return value of one of the tasks that completed.
+    ///  * `Poll::Ready(Err(err))` if one of the tasks in this `TaskSet` has panicked or been
+    ///     aborted.
+    ///  * `Poll::Ready(Ok(None))` if the `TaskSet` is empty.
+    ///
+    /// Note that this method may return `Poll::Pending` even if one of the tasks has completed.
+    /// This can happen if the coop budget is reached.
+    pub fn poll_join_one(&mut self, cx: &mut Context<'_>) -> Poll<Result<Option<T>, JoinError>> {
```

Comment:
I would avoid making this public for now. It *might* make it trickier to implement `AsyncIterator` in the future. I have not thought through this yet though.